### PR TITLE
Try to solve issue3456.

### LINKED ISSUE #3456: format_strings option being ignored for strings with escaped newlines
In my rustfmt.toml file, I have specified:

    format_strings = false

However, when I run rustfmt, I see my escaped strings getting formatted.  The following code:

    println!(
        "DirEntry {{ \
            binary_name: {:<64}, \
            context_id: {:>2}, \
            file_size: {:>6}, \
            offset: 0x{:>08X}, \
            actual_crc: 0x{:>08X} \
        }}",
        dir_entry.binary_name,
        dir_entry.context_id,
        dir_entry.file_size,
        dir_entry.offset,
        dir_entry.actual_crc
    );

Gets formatted to:

    println!(
        "DirEntry {{ \
            binary_name: {:<64}, \
            context_id: {:>2}, \
            file_size: {:>6}, \
            offset: 0x{:>08X}, \
            actual_crc: 0x{:>08X} \
            }}",
        dir_entry.binary_name, dir_entry.context_id, dir_entry.file_size, dir_entry.offset, dir_entry.actual_crc
    );

### LINKED ISSUE #3860: Inconsistent alignment of string constants
rustfmt did this:

```diff
 const SCISSOR_LINE: &str = "\
-# ------------------------ >8 ------------------------";
+                            # ------------------------ >8 ------------------------";
 const SCISSOR_COMMENT: &str = "\
 # Do not touch the line above.
 # Everything below will be removed.
 ";
```

This seems wrong for two reasons: rustfmt shouldn't use visual alignment for anything by default, and rustfmt should not reformat the former string but not the latter.

rustfmt should have left both of these strings untouched.

### PR DESCRIPTION
**Pull request by xiongmao86**
Close #3456.
Close #3860.

### CODE REVIEW

**Comment by xiongmao86** [comment]
I read code and find this snippet relevant:
https://github.com/rust-lang/rustfmt/blob/cc97eaf9f9181aec5a1ba6459b204ed4a7fba62a/src/expr.rs#L1251-L1277

This branch of code indents strings even when config.format_strings() is false.

I git log and find #1724 is the pull request introduce this condition branch, and #1661 is the issues that @1724 solve. Unfortunately, the code reference is not working, so I can't see why that change is make. What can I do? Any suggestion?

**Comment by xiongmao86** [comment]
Oh, multi-line string literal end with \, and '\\' is character \ in a escape form.

**Comment by scampi** [comment]
@xiongmao86 exactly. It checks that every line ends with `\` and if they do, it aligns them. In effect, it doesn't change the string's layout (which is good), but I don't know why it was added based on the issue you shared...
@topecongiro do you remember the reason for this ?

**Comment by xiongmao86** [comment]
There are 4 tests failed in current code. 1 self test, and 3 system tests. 

The self test indent the second line of a string lit that was re-indented but the code before change. So some function that changed the indent of the second line changed back by the changing code snippet.

The system tests pertain to the argument processing of function and macro invocation. The code possibly rely on this changing snippet to make sure the invocation never pass the line limit.

**Comment by xiongmao86** [comment]
Identify snippet that make tests to failed:

https://github.com/rust-lang/rustfmt/blob/cc97eaf9f9181aec5a1ba6459b204ed4a7fba62a/src/overflow.rs#L470-L484

**Comment by xiongmao86** [comment]
It's so hard to trace. What mistake did I make?

**Comment by scampi** [comment]
In the part of the code you changed, there was a gated change. Keeping it resolve the issues:

```diff
--- a/src/expr.rs
+++ b/src/expr.rs
@@ -1250,7 +1250,11 @@ fn rewrite_string_lit(context: &RewriteContext<'_>, span: Span, shape: Shape) ->
     let string_lit = context.snippet(span);

     if !context.config.format_strings() {
-        return valid_str(string_lit.to_owned(), context.config.max_width(), shape);
+        return if context.config.version() == Version::Two {
+            Some(string_lit.to_owned())
+        } else {
+            valid_str(string_lit.to_owned(), context.config.max_width(), shape)
+        };
     }
```

However, since your fix modifies a the formatting of a default option, you need to gate what you did as well. See https://github.com/rust-lang/rustfmt/blob/master/Contributing.md#version-gate-formatting-changes for extra infos.

**Comment by xiongmao86** [comment]
@scampi, I think I have tried that option before, and it came up with new issues. I still don't know why changing in this way and what cause changing the layout of the string. I don't know how to narrow down to the location of relevant code.

**Comment by xiongmao86** [comment]
The lines that are ended with '\\' except the last line will process and return early, otherwise the strings will be processed by rewrite_string, which indent the next line.

**Comment by scampi** [comment]
What issues did you get with the proposed https://github.com/rust-lang/rustfmt/pull/3556#issuecomment-498061990 ?
One thing that needs to be fixed for sure is the versioning of the formatting, gated in https://github.com/rust-lang/rustfmt/pull/3556/files#diff-180bb3ad456cc10517d878fb7795341cL1272 but which got removed in this PR. That gating is necessary for the stability guarantee.

**Comment by xiongmao86** [comment]
Well, I try your suggestion and find the second line of the multiline string returned at the version two condition is indented with 4 more space, I was thinking that additional indent is because that second line was process by `rewrite_string`, but after experiment I realize that the string is return early, and hence is not process by the `rewrite_string` function as opposite to my origin assumption.

I can not find out since the string is return early as what it is and not changed by the `rewrite_string`, what cause the second line to indent with 4 extra space?

**Comment by xiongmao86** [comment]
Formatting in `cargo-fmt/main.rs` is fixed in recent version.

**Comment by xiongmao86** [comment]
@scampi, is the line 365 in this snippet in right indentation?

https://github.com/jonesz/stdsimd/blob/074eadce0a14509d10cc31774477848d38380a1f/coresimd/x86/macros.rs#L360-368

**Comment by xiongmao86** [comment]
Oh, that failed test probably ended up like the old self test in `cargo-fmt/main.rs`, may be something wrong in the macro rewriting.

**Comment by xiongmao86** [comment]
Currently when processing the macro definition, the macro branch block is wrap into a main function by pretending function declaration and brace and by appending a close brace before using `format_snippet()` to format the that main function. After that the formatted function is unwrapped, and the content of the block is chopped back one unit of indention(determine by config). 

I found the second line is not affected when I add more "block_indent()` to the block shape. So I think the problem could probably be missing to chop the second line of the multi-line string for one unit of indention. I will try to narrow down in the following days.

**Comment by xiongmao86** [comment]
Test passed. @scampi can you review my commit?

**Comment by xiongmao86** [comment]
If @scampi is not available, are there anyone have time to review my pr?

File: src/utils.rs [hunk 1]
```diff
@@ -600,9 +600,8 @@ pub(crate) fn trim_left_preserve_layout(
 
 /// Based on the given line, determine if the next line can be indented or not.
 /// This allows to preserve the indentation of multi-line literals.
-pub(crate) fn indent_next_line(kind: FullCodeCharKind, line: &str, config: &Config) -> bool {
+pub(crate) fn indent_next_line(kind: FullCodeCharKind, _line: &str, config: &Config) -> bool {
```
Comment (topecongiro) [inline src/utils.rs]: Nit: if the `line` is not used, then we can remove it from the parameters.

**Pull request merged**